columnflow/util.py

```diff
@@ -530,6 +530,16 @@ def __getattr__(self, attr: str) -> Any:
     def __setattr__(self, attr: str, value: Any) -> None:
         self[attr] = value
 
+    def cautious_update(self, other: dict) -> None:
+        """
+        Updates DotDict, but only if no value is overwritten. Throws error otherwise.
+        """
+        key_replicas = set(self.keys()) & set(other.keys())
+        if not key_replicas:
+            self.update(other)
+        else:
+            raise KeyError(f"'{self.__class__.__name__}' object already has attributes {key_replicas}")
+
```

Comment:
2 things here:

- For the DotDict I think it would be better to keep the amount of "real" attributes as minimal as possible and thus rather move such a functionality into a util function that accepts any dict subtypes.
- Maybe more importantly: I think this way of doing "cautious" updates can be dangerous. IIUC, if any of the keys is already existing in the dict, then none of the other dict items are added. But then, one always needs to know whether certain keys are definitely not present, which is quite implicit and might depend on the situation. As an alternative, I would explicitly check for the presence of a certain key (at the point where the update should be performed), and then add new items depending on the response (see e.g. above)

Reply:
I think the idea here was to have a convenience function that works as a regular `dict.update`, but have it fail if any of the keys to be added is already present in the dictionary. The main benefit of this is to prevent accidentally overwriting a value that was set by a previous function operating on the same `DotDict` (e.g. to prevent style changes from getting lost/clobbered when multiple plot functions are called successively). I'm not sure I would consider it dangerous, it would just fail instead of overwriting, so in a sense it functions as an assertion.

This could also be a util function instead of a member of `DotDict`, or one could avoid a function completely and check the presence of a key explicitly whenever a key is added, but having a function would avoid make the calling code less verbose.

It's true that the **values** are not checked. One could of course extend the implementation to fail not just if a key is already present, but only if the value for that key would also change through the update.

Reply:
I added the value check to the new `dict_add_strict` function :+1: